### ISSUE TITLE
ABI inconsistency after LLVM long double alignment change

### ISSUE THREAD

**Issue by kripken** [emscripten-core/emscripten#14606]
Investigating the error on `test_wprintf` that blocks the LLVM roller, it looks like there is an ABI inconsistency after that change. Specifically, variadic arguments are laid out in memory as before, but the `va_arg` code to read from them has been changed (to assume less padding, it seems).

Here is a reduced testacase from `test_wprintf`:

```cpp
// Copyright 2017 The Emscripten Authors.  All rights reserved.
// Emscripten is available under two separate licenses, the MIT license and the
// University of Illinois/NCSA Open Source License.  Both these licenses can be
// found in the LICENSE file.

#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include <wchar.h>

int main()
{
   wprintf (L"floats: %4.2f %+.0e %E \n", 3.1416, 3.1416, 3.1416);

   return 0;
}
```

If I compile the test before and after the alignment change with WASM=0 (for ease of debugging) then main starts like this:

```js
 var $2 = 0, $3 = 0, $4 = 0, $5 = 0;
 $2 = __stack_pointer - 32 | 0;
 __stack_pointer = $2;
 HEAP32[$2 + 16 >> 2] = 776530087;
 HEAP32[$2 + 20 >> 2] = 1074340351;
 HEAP32[$2 + 8 >> 2] = 776530087;
 HEAP32[$2 + 12 >> 2] = 1074340351;
 HEAP32[$2 >> 2] = 776530087;
 HEAP32[$2 + 4 >> 2] = 1074340351;
 $3 = __stack_pointer - 16 | 0;
 __stack_pointer = $3;
 HEAP32[$3 + 12 >> 2] = $2;
 $1 = HEAP32[281];
 $0 = __stack_pointer - 192 | 0;
 __stack_pointer = $0;
 HEAP32[$0 + 188 >> 2] = $2;
 memset($0 + 144 | 0, 0, 36);
 HEAP32[$0 + 184 >> 2] = HEAP32[$0 + 188 >> 2];
 label$1: {
  if ((wprintf_core(0, $0 + 184 | 0, $0, $0 + 144 | 0) | 0) < 0) {
```
Up to the call to wprintf, it is writing the varargs stuff to the stack. That does not change as the alignment changes.

However, the code to read that written data does change. Here is the source:

```cpp
static void pop_arg_long_double(union arg *arg, va_list *ap)
{
  arg->f = va_arg(*ap, long double);
}
```

And here is the diff:

```diff
 function pop_arg_long_double($0, $1) {
  $0 = $0 | 0;
  $1 = $1 | 0;
  var $2 = 0, $3 = 0, $4 = 0, $5 = 0, $6 = 0, $7 = 0, $8 = 0, $9 = 0, $10 = 0, $11 = 0, $12 = 0;
  $2 = $1;
- $1 = HEAP32[$1 >> 2] + 15 & -16;
+ $1 = HEAP32[$1 >> 2] + 7 & -8;
  HEAP32[$2 >> 2] = $1 + 16;
  $12 = $0;
```

It looks like we still write varargs with alignment/padding of 16, but read with alignment/padding of 8 after the change.

@sbc100 I'm not very familiar with LLVM here, but maybe there is another setting we need to flip in clang, to adjust how long double behaves in varargs?

**Comment by kripken** [comment]
Ok, after much poking around through the compiled wasm and lots of printf-debugging, here is what happens:

1. Musl wprintf does this internally: `fprintf("%*.*Lf", x, y, L);`. That passes in the width, precision, and the actual long double (it's a long double because wprintf uses them all the time - we optimized fprintf to not do that, but not wprintf. But that's not a bug, just a code size issue for wprintf users...)
2. The actual code for that fprintf does this: at the address the va_arg stuff is, it puts the first int, then after 4 bytes the second int, **then it skips 8 bytes** and then it places the long double. So the long double is at offset va_arg+16.
3. The fprintf code to read that data, however, does this: twice it does `va_arg(int)`, each time bumping the pointer by 4. That's it. Then the pointer is at offset va_arg+8.
4. Now we try to pop the long double. `va_arg(long double)` aligns the pointer. When we had 16-byte alignment, it would add 8 to make the concrete value aligned... which is exactly the right address, at va_arg+16. However, with the lower alignment of 8, now we only ensure it is 8-byte aligned, which it already is actually. So we do nothing, and try to read the long double from address va_arg+8, which is wrong.

In conclusion, the varargs writing code appears to be adding padding to ensure 16-byte alignment of long doubles. I think this might be a bug in the varargs lowering code in LLVM? Maybe it has a hardcoded 16 in there somehow that we need to update to 8?

**Comment by dschuff** [comment]
OK, so the problem is that the vararg argument lowering is ultimately driven by the datalayout for the `f128` LLVM type, rather than clang's EmscriptenTargetInfo. If we add `f128:64' to the datalayout string (both in `WebAssembly32TargetInfo` in clang/lib/Basic/Targets/WebAssembly.h and in `WebAssemblyTargetMachine` on the LLVM side in llvm/lib/Target/WebAssemblyTargetmachine.cpp, because they are duplicated for some reason), then it seems to work. That makes it more annoying to differentiate between emscripten and non-emscripten, I'll have to see how difficult that is.

**Comment by dschuff** [comment]
https://reviews.llvm.org/D105749 has the fix

**Comment by kripken** [comment]
Fixed by that patch landing in LLVM.
